enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,467 @@
+      - `query`: Specifies the Prometheus query to use.
+      - `authModes`: Specifies the authentication method to use. Should support
+        at least *basic*, *bearer* and *tls* authentication.
+      - `ignoreNullValues`: false 
+      - `unsafeSsl`: Specifies whether the certificate check should be skipped.
+    - `cpu`:
+      - `value`: Load average value that determines if scaling has to occur.
+      - `timeWindow`: Duration over which the load average must be higher to
+        trigger a scale up.
+    - `memory`:
+      - `value`: Load average value that determines if scaling has to occur.
+      - `timeWindow`: Duration over which the load average must be higher to
+        trigger a scale up.
+- `scaleDown`: defines configuration for scaling up. This includes the following
+  sub elements:
+  - `stabilizationWindow`: duration to wait after a scale down before another
+    scaling operation can occur.
+  - `selectPolicy`: `next` to select the next smaller instance size.
+  - `triggerPolicy`: `all` or `any`. Determines if scaling occurs if one or
+    all triggers have to be true.
+  - `triggers`: sets up the list of triggers, which will trigger a scale down.
+    For details see the `scaleUp` configuration.
+    An empty list of triggers disables scale down.
+
+A complete custom resources instance in YAML format could then look like this:
+
+```yaml
+controlplaneautoscaling:
+  machineConfiguration:
+    machineType: machines_v1beta1_machine_openshift_io
+    machines_v1beta1_machine_openshift_io:
+      machineConfigurations:
+        platform: AWS
+        aws:
+        - weight: 1
+          instanceSize: r4.large
+        - weight: 2
+          instanceSize: r4.xlarge
+  syncPeriod: [duration as specified in golang: e.g. 5s, 15s]
+  scaleUp:
+    selectPolicy: "next"
+    stabilizationWindow: [duration as specified in golang: e.g. 5s, 15s]
+    triggerPolicy: "any"
+    triggers:
+      - type: "cpu"
+        value: "80"
+        timeWindow: "30m"
+  scaleDown:
+    selectPolicy: "next"
+    stabilizationWindow: [duration as specified in golang: e.g. 5s, 15s]
+    triggerPolicy: "all"
+    triggers:
+      - type: "cpu"
+        value: "80"
+        timeWindow: "30m"
+      - type: "memory"
+        value: "90"
+        timeWindow: "30m"
+```
+
+In case the user decides to use `prometheus`, authentication will be specified
+in an additional custom resource
+`ControlPlaneMachineSetAutoscalingTriggerAuthentication` that is based on
+[CMA](https://docs.openshift.com/container-platform/4.13/nodes/cma/nodes-cma-autoscaling-custom-trigger.html#nodes-cma-autoscaling-custom-prometheus-config_nodes-cma-autoscaling-custom-trigger
+) as well and specifies the following values:
+
+- `secretTargetRef`:
+  - `parameter`: type of the secret referenced: should be `bearer` for bearer
+    authentication.
+  - `name`: name of the secret too use.
+  - `key`: key in the secret that contains the token.
```

Comment:
Why not include this directly? If you do include a separate CR for this (not an uncommon pattern, seen that before), I would at least expect that there is a reference to the authentication CR in the prometheus trigger struct

Reply:
I based this on https://keda.sh/docs/2.12/scalers/prometheus/#authentication-parameters - same as the trigger configuration for consistency, but I can also move it directly into the trigger configuration.